src/v/storage/tests/compaction_e2e_test.cc

```diff
@@ -526,3 +627,554 @@ TEST_F(CompactionFixtureTest, TestReadFilledGapsWithTerms) {
+        ASSERT_NO_FATAL_FAILURE();
+
+        ASSERT_EQ(consumed_kvs.size(), latest_kv_map.size());
+
+        // Assert the key consumed is in the latest_kv_map.
+        for (const auto& kv : consumed_kvs) {
+            ASSERT_TRUE(latest_kv_map.contains(kv.key));
+            ASSERT_EQ(kv.val, latest_kv_map[kv.key]);
+        }
+    }
+
+    // Maybe sleep for tombstone.retention.ms time, so that the next time we
+    // attempt to compact the tombstone records will be eligible for deletion.
+    if (wait_for_retention_ms) {
+        ss::sleep(tombstone_retention_ms.value()).get();
+    }
+
+    // Generate one record, so that sliding window compaction can occur.
+    const auto num_extra_records = 1;
+    generate_data(
+      1, cardinality * 2, 1, 1, cardinality + 1, false, &latest_kv_map)
+      .get();
+    did_compact = do_sliding_window_compact(
+                    log->segments().back()->offsets().get_base_offset(),
+                    tombstone_retention_ms)
+                    .get();
+
+    ASSERT_TRUE(did_compact);
+    {
+        tests::kafka_consume_transport consumer(make_kafka_client().get());
+        consumer.start().get();
+        auto consumed_kvs = consumer
+                              .consume_from_partition(
+                                topic_name,
+                                model::partition_id(0),
+                                model::offset(0))
+                              .get();
+
+        // Assert the key consumed is in the latest_kv_map.
+        // latest_kv_map.size() != consumed_kvs.size() if we waited for
+        // retention_ms, due to tombstone removal.
+        for (const auto& kv : consumed_kvs) {
+            ASSERT_TRUE(latest_kv_map.contains(kv.key));
+            ASSERT_EQ(kv.val, latest_kv_map[kv.key]);
+        }
+
+        if (wait_for_retention_ms) {
+            // Assert there is no tombstone record left.
+            auto find_tombstone_record = [](const auto& kv) {
+                return kv.is_tombstone;
+            };
+
+            auto tombstone_it = std::find_if(
+              consumed_kvs.begin(), consumed_kvs.end(), find_tombstone_record);
+
+            // The tombstones should have been removed after second round of
+            // compaction post tombstone.retention.ms.
+            ASSERT_EQ(tombstone_it, consumed_kvs.end());
+            ASSERT_EQ(
+              consumed_kvs.size(), num_records_produced + num_extra_records);
+        } else {
+            ASSERT_EQ(consumed_kvs.size(), latest_kv_map.size());
+            ASSERT_EQ(
+              consumed_kvs.size(),
+              num_tombstones_produced + num_records_produced
+                + num_extra_records);
+        }
+    }
+}
+
+INSTANTIATE_TEST_SUITE_P(
+  RandomDistribution,
+  CompactionFixtureTombstonesRandomParamTest,
+  ::testing::Combine(
+    ::testing::Values(
+      TombstonesRandomArgs::create(),
+      TombstonesRandomArgs::create(),
+      TombstonesRandomArgs::create()),
```

Comment:
nit: this seems a bit of an odd use of parameterized testing. Would kind of expect params to be deterministic, and for random generation to be done in test. Otherwise the test name becomes non-deterministic

I suggest just having the test generate the args, and running with those args just once, and rely on test repeats to expand coverage

Reply:
Changed to parameterize test with only one instance of `TombstonesRandomArgs::create()`.